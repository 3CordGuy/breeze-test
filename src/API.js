const getGroups = () =>
    fetch(`//localhost:8000/api/groups/`).then((response) => response.json());

const getPeople = () =>
    fetch(`//localhost:8000/api/people/`).then((response) => response.json());

const removePerson = (person_id) =>
    fetch(`//localhost:8000/api/people/${person_id}`, {
        method: "DELETE",
        headers: {
            "Content-Type": "application/json",
        },
    });

const updatePerson = (person) =>
    fetch(`//localhost:8000/api/people/${person.id}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(person),
    });

const importPeople = (people) =>
    fetch(`//localhost:8000/api/people-import`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(people),
    });

const importGroups = (groups) =>
    fetch(`//localhost:8000/api/groups-import`, {
        method: "POST",
        mode: "cors",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(groups),
    });

const getGroup = (group_id) =>
    fetch(`http://localhost:8000/api/groups/${group_id}`).then((response) =>
        response.json(),
    );

const API = {
    getGroups,
    getPeople,
    removePerson,
    updatePerson,
    importPeople,
    importGroups,
    getGroup,
};

export default API;
