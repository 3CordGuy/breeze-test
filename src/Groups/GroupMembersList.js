import React, { Component } from "react";
import { Table, Header, Button, Icon } from "semantic-ui-react";
import _ from "lodash";

class GroupMembersList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            group: "Group",
            members: [],
            column: "last_name",
            direction: "descending",
        };
    }

    handleSort = (clickedColumn) => () => {
        const { column, members, direction } = this.state;

        if (column !== clickedColumn) {
            this.setState({
                column: clickedColumn,
                members: _.sortBy(members, [clickedColumn]),
                direction: "ascending",
            });

            return;
        }

        this.setState({
            members: members.reverse(),
            direction: direction === "ascending" ? "descending" : "ascending",
        });
    };

    componentDidMount() {
        let { group_id } = this.props;
        fetch(`http://localhost:8000/api/groups/${group_id}`)
            .then((response) => response.json())
            .then(({ data }) =>
                this.setState({
                    members: _.sortBy(data.members, [
                        "last_name",
                        "descending",
                    ]),
                    group: data.name,
                }),
            );
    }

    render() {
        let { column, direction, members, group } = this.state;

        return (
            <>
                <Header as="h1">{group}</Header>
                <Table celled padded sortable>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell singleLine>
                                First Name
                            </Table.HeaderCell>
                            <Table.HeaderCell
                                singleLine
                                sorted={
                                    column === "last_name" ? direction : null
                                }
                                onClick={this.handleSort("last_name")}
                            >
                                Last Name
                            </Table.HeaderCell>
                            <Table.HeaderCell>Email</Table.HeaderCell>
                            <Table.HeaderCell>Status</Table.HeaderCell>
                            <Table.HeaderCell></Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        {members.map((person, index) => {
                            return (
                                <Table.Row key={index}>
                                    <Table.Cell singleLine>
                                        {person.first_name}
                                    </Table.Cell>
                                    <Table.Cell singleLine>
                                        {person.last_name}
                                    </Table.Cell>
                                    <Table.Cell singleLine>
                                        {person.email_address}
                                    </Table.Cell>
                                    <Table.Cell singleLine>
                                        {person.status}
                                    </Table.Cell>
                                    <Table.Cell singleLine textAlign="center">
                                        <Button icon circular primary>
                                            <Icon name="edit" />
                                        </Button>
                                        <Button icon negative basic circular>
                                            <Icon name="trash" />
                                        </Button>
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })}
                    </Table.Body>
                </Table>
            </>
        );
    }
}

export default GroupMembersList;
