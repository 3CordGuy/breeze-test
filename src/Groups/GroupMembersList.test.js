import React from "react";
import { shallow, mount } from "enzyme";
import GroupMembersList from "./GroupMembersList";
import API from "../API";
import _ from "lodash";

let MEMBERS_RESPONSE = {
    data: {
        id: 62,
        name: "Pawnee Book Club",
        created_at: "2020-06-28T01:17:03.000000Z",
        updated_at: "2020-06-28T01:17:03.000000Z",
        members: [
            {
                id: 128,
                first_name: "Leslie",
                last_name: "Knope",
                email_address: "leslise@pawnee.in.gov",
                status: "active",
                created_at: "2020-06-28T15:12:07.000000Z",
                group: {
                    id: 62,
                    name: "Pawnee Book Club",
                    created_at: "2020-06-28T01:17:03.000000Z",
                    updated_at: "2020-06-28T01:17:03.000000Z",
                },
                updated_at: "2020-06-28T15:12:07.000000Z",
            },
            {
                id: 132,
                first_name: "Wilma",
                last_name: "Flintstone",
                email_address: "wilmaflinstone@example.com",
                status: "active",
                created_at: "2020-06-28T15:12:07.000000Z",
                group: {
                    id: 62,
                    name: "Pawnee Book Club",
                    created_at: "2020-06-28T01:17:03.000000Z",
                    updated_at: "2020-06-28T01:17:03.000000Z",
                },
                updated_at: "2020-06-28T15:12:07.000000Z",
            },
        ],
    },
};

let wrapper, data;

// SNAPSHOT TEST
describe("<GroupMembersList />", () => {
    beforeAll(() => {
        wrapper = shallow(<GroupMembersList />);

        data = [
            {
                id: 132,
                name: "Awesome Group",
                members: [
                    {
                        id: 132,
                        first_name: "Macie",
                        last_name: "Emmerich",
                        email_address: "cremin.marjory@hotmail.com",
                        status: "active",
                        updated_at: "2019-07-20 22:05:47",
                        created_at: "2019-07-20 22:05:47",
                    },
                ],
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
        ];

        wrapper.setState({ group: data[0].name, members: data[0].members });
    });

    test("should match the snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
});

// // UNIT TESTS
describe("<GroupMembersList />", () => {
    it("should show a proper loader", () => {
        const wrapper = mount(<GroupMembersList />);

        wrapper.setState({ loading: true });

        let loader = wrapper.find("div.loader");

        expect(loader.text()).toBe("Loading Data...");
    });

    it("should fetch the groups async", () => {
        const promise = new Promise((resolve) => resolve(MEMBERS_RESPONSE));

        API.getGroup = jest.fn(() => promise);

        const wrapper = mount(<GroupMembersList />);

        expect(wrapper.find("table tbody tr").length).toEqual(1);

        promise.then(() => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find("table tbody tr").length).toEqual(2);

                API.getGroup.mockClear();
            });
        });
    });
});

// INTEGRATION

describe("<GroupMembersList />", () => {
    it("should sort list by group name", () => {
        const wrapper = mount(<GroupMembersList />);

        wrapper.setState({
            direction: "asc",
            data: _.orderBy(MEMBERS_RESPONSE.data, "name", ["asc"]),
        });

        let nameHeader = wrapper.find("th.ascending.sorted");

        nameHeader.simulate("click");
        wrapper.update();

        expect(wrapper.state().data).toEqual(
            _.orderBy(MEMBERS_RESPONSE.data, "name", ["desc"]),
        );

        nameHeader.simulate("click");

        expect(wrapper.state().data).toEqual(
            _.orderBy(wrapper.state().data, "name", ["asc"]),
        );
    });

    it("should warn when trying to remove a person", () => {
        const wrapper = mount(<GroupMembersList />);
        window.confirm = jest.fn();

        wrapper.setState({
            members: [
                {
                    id: 132,
                    first_name: "Jim",
                    last_name: "Halpert",
                    email_address: "jim.halpert@dundermifflin.com",
                    status: "active",
                    updated_at: "2019-07-20 22:05:47",
                    created_at: "2019-07-20 22:05:47",
                },
            ],
        });

        let trashButton = wrapper.find("button.negative");
        trashButton.simulate("click");

        expect(window.confirm).toBeCalled();
    });

    it("should allow removal of a group member", async () => {
        const wrapper = mount(<GroupMembersList />);

        wrapper.setState({
            direction: "asc",
            data: _.orderBy(MEMBERS_RESPONSE.data, "name", ["asc"]),
        });

        const removePersonPromise = new Promise((resolve) => {
            let newMembersList = MEMBERS_RESPONSE.data;
            newMembersList.members = newMembersList.members.filter(
                (member) => member.id === 128,
            );
            return resolve(newMembersList);
        });

        API.removePerson = jest.fn(() => removePersonPromise);

        await removePersonPromise;

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find("table tbody tr").length).toEqual(1);

            API.removePerson.mockClear();
        });
    });
});
