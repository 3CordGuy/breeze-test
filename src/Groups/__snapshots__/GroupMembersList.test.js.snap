// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupMembersList /> should match the snapshot 1`] = `
<Fragment>
  <Dimmer
    active={true}
    inverted={true}
  >
    <Loader>
      Loading Data...
    </Loader>
  </Dimmer>
  <Button
    as="button"
    onClick={[Function]}
  >
    Back
  </Button>
  <Header
    as="h1"
  >
    Awesome Group
     
    <HeaderSubheader>
      Group Members
    </HeaderSubheader>
  </Header>
  <Table
    as="table"
    basic="very"
    celled={true}
    padded={true}
    sortable={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          singleLine={true}
          sorted={null}
        >
          First Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          singleLine={true}
          sorted="descending"
        >
          Last Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted={null}
        >
          Email
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
        >
          Status
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
        />
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="0"
      >
        <TableCell
          as="td"
          singleLine={true}
        >
          Macie
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          Emmerich
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          cremin.marjory@hotmail.com
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          active
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
          textAlign="center"
        >
          <Button
            as="button"
            circular={true}
            icon={true}
            primary={true}
          >
            <Icon
              as="i"
              name="edit"
            />
          </Button>
          <Button
            as="button"
            basic={true}
            circular={true}
            icon={true}
            negative={true}
          >
            <Icon
              as="i"
              name="trash"
            />
          </Button>
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
</Fragment>
`;
